File: command-center/ui/pages/observatory.js
/* Six read-only projections over canonical records, with no model/brain fiction. */
import {api} from '../api.js';
import {h,toastError} from '../components.js';
import {setTraceEnabled,setTraceRoutes,traceSnapshot,clearTrace} from '../api_trace.js';
const VIEWS={brain:'Brain View',agents:'Live Agent Graph',memory:'Memory Map',learning:'Learning View',xray:'UI → API X-Ray',analysis:'Self-Analysis'};
const text=v=>v===null||v===undefined?'UNKNOWN':String(v);
const tag=v=>h('span.obs-tag',text(v));
const note=t=>h('p.small.dim',t);
const panel=(title,...children)=>h('section.panel.obs-panel',h('h3',title),...children);
const button=(label,fn)=>h('button.btn.btn-sm',{type:'button',onClick:fn},label);
const empty=()=>note('Нет записей в выбранной области. Это не успешный тест и не подтверждение отсутствия проблем.');
function table(headers,rows){
  if(!rows.length)return empty();
  return h('div.obs-table',h('table',h('thead',h('tr',...headers.map(x=>h('th',x)))),h('tbody',...rows.map(row=>h('tr',...row.map(x=>h('td',typeof x==='object'&&x!==null&&x.nodeType?x:text(x))))))));
}
function styles(){
  if(document.getElementById('observatory-style'))return;
  document.head.append(h('link',{id:'observatory-style',rel:'stylesheet',href:'/pages/observatory.css'}));
}
export function graphData(data,mode){
  const nodes=new Map(),edges=[];
  const add=(id,label,level,status)=>{if(!nodes.has(id))nodes.set(id,{id,label,level,status});};
  const edge=(from,to)=>{if(nodes.has(from)&&nodes.has(to))edges.push({from,to});};
  if(mode==='memory'){
    for(const f of data.memory.facts){add('f'+f.id,'Fact #'+f.id,1,f.invalid_at||f.expired_at?'HISTORICAL':'RECORDED');if(f.source_run_id!=null)add('r'+f.source_run_id,'Run #'+f.source_run_id,0,'SOURCE_ID');}
    for(const f of data.memory.facts){if(f.source_run_id!=null)edge('r'+f.source_run_id,'f'+f.id);if(f.superseded_by!=null)edge('f'+f.id,'f'+f.superseded_by);}
  }else{
    for(const t of data.tasks){if(t.agent_id!=null)add('a'+t.agent_id,'Agent #'+t.agent_id,0,'ASSIGNED');add('t'+t.id,'Task #'+t.id,1,t.status);if(t.agent_id!=null)edge('a'+t.agent_id,'t'+t.id);}
    for(const r of data.brain.runs){add('t'+r.task_id,'Task #'+r.task_id,1,'SELECTED');add('r'+r.id,'Run #'+r.id,2,r.status);edge('t'+r.task_id,'r'+r.id);if(r.model_id!=null){add('m'+r.model_id,'Model #'+r.model_id,3,r.model_kind);edge('r'+r.id,'m'+r.model_id);}}
    for(const c of data.brain.tool_calls){add('c'+c.id,c.tool+' #'+c.id,3,c.status);edge('r'+c.run_id,'c'+c.id);}
  }
  const visible=[...nodes.values()].slice(0,100), ids=new Set(visible.map(n=>n.id));
  return {nodes:visible,edges:edges.filter(e=>ids.has(e.from)&&ids.has(e.to)),total:nodes.size};
}
function graph(data,mode,ctx,params){
  const p=graphData(data,mode);if(!p.nodes.length)return empty();
  const counts={};let rows=0;const pos=new Map();
  p.nodes.forEach(n=>{const row=counts[n.level]||0;counts[n.level]=row+1;rows=Math.max(rows,row+1);pos.set(n.id,{x:20+n.level*250,y:25+row*66});});
  const ns='http://www.w3.org/2000/svg';
  const svg=document.createElementNS(ns,'svg');svg.setAttribute('viewBox',`0 0 1030 ${Math.max(170,rows*66+40)}`);svg.setAttribute('role','img');svg.setAttribute('aria-label',mode==='memory'?'Recorded memory lineage':'Recorded agent task run graph');
  const make=(type,attrs={})=>{const n=document.createElementNS(ns,type);for(const [k,v]of Object.entries(attrs))n.setAttribute(k,String(v));return n;};
  for(const e of p.edges){const a=pos.get(e.from),b=pos.get(e.to);svg.append(make('path',{d:`M${a.x+205},${a.y+23} C${a.x+238},${a.y+23} ${b.x-28},${b.y+23} ${b.x},${b.y+23}`,class:'obs-edge'}));}
  for(const n of p.nodes){const point=pos.get(n.id);const g=make('g',{'data-node-id':n.id});g.append(make('rect',{x:point.x,y:point.y,width:205,height:48,rx:8,class:'obs-node'}));
    const label=make('text',{x:point.x+9,y:point.y+18,class:'obs-label'});label.textContent=n.label.slice(0,28);g.append(label);
    const status=make('text',{x:point.x+9,y:point.y+37,class:'obs-status'});status.textContent=n.status;g.append(status);
    if(n.id.startsWith('t')){g.setAttribute('tabindex','0');g.setAttribute('role','button');g.setAttribute('aria-label','Открыть задачу '+n.id.slice(1));const open=()=>ctx.navigate('observatory',{...params,task:n.id.slice(1),view:'brain'});g.addEventListener('click',open);g.addEventListener('keydown',e=>{if(e.key==='Enter')open();});}
    svg.append(g);
  }
  return h('div.obs-graph',note(`Узлы ${p.nodes.length}/${p.total}. Линии отражают только сохранённые ID-связи, не сходство мыслей.`),svg);
}
function brain(d){const b=d.brain;return h('div.stack',
  panel('Задача → запуск → модель → инструмент',note('Показаны события и метаданные выполнения, не скрытые рассуждения модели. Запись completed не доказывает независимую проверку результата.'),
    table(['Run','Состояние','Модель','Вход / выход, токены','Стоимость в журнале, $'],b.runs.map(r=>[r.id,tag(r.status),r.model_id==null?'UNKNOWN':`${r.model_kind} #${r.model_id}`,`${text(r.tokens_in)} / ${text(r.tokens_out)}`,r.cost_usd]))),
  panel('Инструменты и права',table(['Вызов','Run / шаг','Инструмент','Режим','Состояние','мс'],b.tool_calls.map(c=>[c.id,`${c.run_id} / ${c.step}`,c.tool,tag(c.effect),tag(c.status),c.duration_ms]))),
  panel('Подтверждения',table(['ID','Run','Статус','Создано'],b.approvals.map(a=>[a.id,a.run_id,tag(a.status),a.created_at]))),
  panel('Записанные проверки',note('Это записи verification.result, а не повторная независимая проверка этой панелью.'),table(['ID','Run','Записанный статус','Время'],b.recorded_verifications.map(v=>[v.id,v.run_id,tag(v.status),v.ts]))));}
function memory(d,ctx,p){return panel('Память: происхождение и замены',note(`Область: ${d.memory.scope}. Только метаданные; тексты памяти и личные файлы не читаются этим представлением.`),graph(d,'memory',ctx,p),table(['Fact','Источник','Run','Заменён на','Уверенность в записи'],d.memory.facts.map(f=>[f.id,f.source_kind,f.source_run_id,f.superseded_outside_window?'OUTSIDE_WINDOW':f.superseded_by,f.confidence])));}
function learning(d){return panel('Сравнение версий навыков',note('Последние 50 записей оценок экземпляра. Рекомендация promote не означает применение. Здесь нет кнопки обучения весов, установки обновления или изменения разрешений.'),
 table(['Оценка / навык','Версии: база → кандидат','Состояние / вердикт','Применено','Запуски','Успешность: база → кандидат','мс: база → кандидат'],d.learning.evaluations.map(e=>[`${e.id} / ${e.skill_id}`,`${e.baseline_version_id} → ${e.candidate_version_id}`,`${e.status} / ${e.verdict}`,e.applied===true?'YES':'NO',`${text(e.baseline_runs)} → ${text(e.candidate_runs)}`,`${text(e.baseline_success_rate)} → ${text(e.candidate_success_rate)}`,`${text(e.baseline_avg_duration_ms)} → ${text(e.candidate_avg_duration_ms)}`])));}
function xray(ctx){
  const area=h('div'),info=note('');
  const paint=()=>{const s=traceSnapshot();info.textContent=`${s.enabled?'ON':'OFF'} · ${s.rows.length}/${s.limit} записей только этой вкладки. Raw fetch, WebSocket, серверные и DB spans не перехватываются.`;
    area.replaceChildren(table(['UI-метка','Метод','Контракт маршрута','HTTP','мс','Транспорт'],s.rows.map(r=>[r.ui_action,r.method,r.path,r.status,r.duration_ms,r.outcome])));};
  paint();
  return panel('UI → API X-Ray',note('Запись выключена по умолчанию. URL-параметры, тела, заголовки, ответы и секреты не сохраняются. При logout данные стираются. HTTP_OK означает транспорт, не успешную бизнес-операцию.'),
    h('div.row',button('Включить X-Ray',async()=>{try{const routes=await api.raw('/api/observatory/routes');setTraceRoutes(routes.routes);setTraceEnabled(true);paint();}catch(e){toastError(e);}}),button('Выключить и очистить',()=>{setTraceEnabled(false);paint();}),button('Обновить X-Ray',paint),button('Очистить',()=>{clearTrace();paint();})),info,area);
}
function analysis(d){const names={FAILED_TASKS:'Задачи со статусом failed в последних 50',WAITING_APPROVAL:'Ждут подтверждения в последних 50',COMPLETED_WITHOUT_VERIFICATION_RECORD_IN_WINDOW:'Завершённые запуски выбранной задачи без записи проверки в окне',TOOL_ERROR_IN_SELECTED_WINDOW:'Ошибки инструментов выбранной задачи в окне'};
 return panel('Самоанализ по наблюдаемым данным',note('Детерминированная диагностика, не оценка IQ и не самостоятельное изменение ядра. Усечённая выборка не доказывает отсутствие проблем.'),
 h('div.obs-cards',...d.analysis.observations.map(o=>h('article.obs-card',h('strong.obs-number',text(o.count)),h('span',names[o.code]||o.code)))),note('Восстановление: откройте задачу и журнал, проверьте её права и состояние. Не повторяйте внешний эффект с неизвестным исходом. Автоисправления отключены.'));
}
export default {
 id:'observatory',title:'Brain & Observatory',icon:'activity',nav:'primary',section:'brains',
 async render(ctx,params={}){
  styles();const view=VIEWS[params.view]?params.view:'brain';const query=new URLSearchParams();
  if(params.task)query.set('task_id',params.task);if(params.project)query.set('project_id',params.project);
  const root=h('div.obs-root.stack.lg',h('div',h('p.small.dim','BOSSMAN / OBSERVATORY'),h('h1','Система под наблюдением'),note('Шесть представлений одного ядра. Никаких фиктивных нейронов, демо-телеметрии или нового агента.')));
  root.append(h('nav.obs-tabs',...Object.entries(VIEWS).map(([id,label])=>h('button.btn',{type:'button','aria-pressed':id===view,onClick:()=>ctx.navigate('observatory',{...params,view:id})},label))));
  try{
   const d=await api.raw('/api/observatory/snapshot'+(query.size?'?'+query:''),{traceOrigin:'Observatory.snapshot'});
   if(d.schema_version!==1||!Array.isArray(d.tasks)||!d.brain)throw Error('Неподдерживаемый ответ Observatory.');
   const select=h('select.input',{'aria-label':'Задача Observatory',onChange:e=>ctx.navigate('observatory',{...params,task:e.target.value,view})},h('option',{value:''},'Выберите задачу'),...d.tasks.map(t=>h('option',{value:String(t.id)},`#${t.id} · ${t.status}`)));
   if(params.task&&!d.tasks.some(t=>String(t.id)===params.task))select.append(h('option',{value:params.task},'#'+params.task+' · вне окна задач'));
   select.value=params.task||'';
   const project=h('input.input',{'aria-label':'Проект Memory Map',value:params.project||'',placeholder:'ID проекта для карты памяти',maxLength:96});
   root.append(h('div.obs-controls',select,project,button('Открыть область памяти',()=>ctx.navigate('observatory',{...params,project:project.value.trim(),view:'memory'})),button('Обновить данные',()=>ctx.refresh())),
    note(`Источник: ${d.source} · снимок ${d.observed_at} · WebSocket: ${ctx.bus?.state||'UNKNOWN'}. Обновления только при событиях открытой страницы или вручную.`));
   if(Object.values(d.truncated).some(Boolean)||d.memory.truncated||d.learning.truncated)root.append(h('p.obs-warning','Выборка ограничена. Показанные счётчики и отсутствие записей относятся только к этому окну.'));
   root.append(view==='brain'?brain(d):view==='agents'?panel('Агенты → задачи → запуски → инструменты',graph(d,'agents',ctx,params)):view==='memory'?memory(d,ctx,params):view==='learning'?learning(d):view==='xray'?xray(ctx):analysis(d));
  }catch(e){root.append(panel('Данные недоступны',h('p',{role:'alert'},e.message),button('Повторить чтение',()=>ctx.refresh())));}
  return root;
 },
 onEvent(ev){return document.visibilityState==='visible'&&/^(task\.|run\.|tool\.|approval\.|fact\.|skill\.|verification\.|ws\.)/.test(String(ev.kind||''));}
};

File: command-center/bcc/features/observatory.py
"""Owner-only read projections over existing BCC records. No second store.

Never select prompts, reasoning/checkpoints, raw results, errors, tokens, keys,
request bodies, personal filenames or memory content. Task/project scopes are
explicit. An execution record is NOT independent verification.
"""
from __future__ import annotations

import asyncio
from collections import Counter
from datetime import datetime, timezone
import math

import sqlalchemy as sa
from fastapi import APIRouter, HTTPException, Query, Request

from . import Feature
from .. import db
from ..tools import REGISTRY

router = APIRouter()
LIMIT = 100


def _enum(value, allowed):
    return value if isinstance(value, str) and value in allowed else "UNKNOWN"


def _finite(value):
    return value if type(value) in (int, float) and math.isfinite(value) and value >= 0 else None


async def _rows(session, table, columns, where=None, limit=LIMIT):
    stmt = sa.select(*(table.c[c] for c in columns)).order_by(table.c.id.desc()).limit(limit + 1)
    if where is not None:
        stmt = stmt.where(where)
    rows = [dict(row) for row in (await session.execute(stmt)).mappings()]
    return rows[:limit], len(rows) > limit


async def snapshot(svc, task_id=None, project_id=None):
    async with svc.db.session() as s:
        tasks, tasks_cut = await _rows(s, db.tasks, ["id", "agent_id", "status", "updated_at"], limit=50)
        for row in tasks:
            row["status"] = _enum(row["status"], {"draft","queued","running","paused","waiting_approval","completed","failed","stopped"})
        runs, calls, approvals, records, verifications = [], [], [], [], []
        truncation = {"tasks": tasks_cut}
        if task_id is not None:
            if not (await s.execute(sa.select(db.tasks.c.id).where(db.tasks.c.id == task_id))).first():
                raise HTTPException(404, "Task not found")
            # Join model identity, never display arbitrary model_alias or secret fields.
            result = await s.execute(sa.select(db.task_runs.c.id, db.task_runs.c.task_id,
                db.task_runs.c.status, db.task_runs.c.started_at, db.task_runs.c.finished_at,
                db.task_runs.c.tokens_in, db.task_runs.c.tokens_out, db.task_runs.c.cost_usd,
                db.models.c.id.label("model_id"), db.models.c.kind.label("model_kind"))
                .select_from(db.task_runs.outerjoin(db.models, db.models.c.alias == db.task_runs.c.model_alias))
                .where(db.task_runs.c.task_id == task_id).order_by(db.task_runs.c.id.desc()).limit(51))
            all_runs = [dict(r) for r in result.mappings()]
            runs, truncation["runs"] = all_runs[:50], len(all_runs)>50
            run_ids = [r["id"] for r in runs]
            for r in runs:
                r["status"] = _enum(r["status"], {"queued","leased","running","completed","failed","stopped"})
                r["model_kind"] = _enum(r["model_kind"], {"local","cloud"})
                # These are ledger fields, not proof of a live model/provider call.
                for key in ("tokens_in","tokens_out","cost_usd"):
                    r[key] = _finite(r[key])
            calls, truncation["calls"] = await _rows(s, db.tool_calls,
                ["id","run_id","step","tool","effect","status","duration_ms"],
                db.tool_calls.c.run_id.in_(run_ids))
            for c in calls:
                # A database string is not a safe display label. Only known registered names.
                c["tool"] = c["tool"] if c["tool"] in REGISTRY.names() else "UNREGISTERED_TOOL"
                c["effect"] = _enum(c["effect"], {"auto","ask","deny"})
                c["status"] = _enum(c["status"], {"pending_approval","approved","rejected","executed","denied","error","timeout","replayed","started","interrupted","reconciled"})
                c["duration_ms"] = _finite(c["duration_ms"])
            approvals, truncation["approvals"] = await _rows(s, db.approvals,
                ["id","run_id","status","created_at","decided_at"], db.approvals.c.task_id == task_id)
            for a in approvals:
                a["status"] = _enum(a["status"], {"pending","approved","rejected","expired","consumed","revoked"})
            # Event kind only: don't interpret free-form event messages as verified outcomes.
            records, truncation["events"] = await _rows(s, db.run_events,
                ["id","run_id","kind","level","ts"], db.run_events.c.run_id.in_(run_ids))
            safe_kinds = {"verification","verification.result","tool.call","tool.result","plan","error","warning","step","model.call","model.result","started","completed","failed"}
            for r in records:
                r["kind"] = _enum(r["kind"], safe_kinds)
                r["level"] = _enum(r["level"], {"info","warn","error","debug"})
            proof = await s.execute(sa.select(db.events.c.id, db.events.c.ts,
                db.events.c.data["run_id"].as_integer().label("run_id"),
                db.events.c.data["status"].as_string().label("status"))
                .where(db.events.c.kind == "verification.result",
                       db.events.c.data["run_id"].as_integer().in_(run_ids))
                .order_by(db.events.c.id.desc()).limit(101))
            all_proof = [dict(r) for r in proof.mappings()]
            verifications, truncation["verifications"] = all_proof[:100], len(all_proof)>100
            for r in verifications:
                r["status"] = _enum(r["status"], {"VERIFIED","UNVERIFIED","FAILED","INCONCLUSIVE"})
        facts, facts_cut = [], False
        memory_scope = "NOT_SELECTED"
        if project_id is not None:
            memory_scope = "EXACT_META_PROJECT_ID"
            where = db.facts.c.meta["project_id"].as_string() == project_id
        elif task_id is not None:
            memory_scope = "TASK_RUN_LINEAGE"
            where = db.facts.c.source_run_id.in_(sa.select(db.task_runs.c.id).where(db.task_runs.c.task_id == task_id))
        if memory_scope != "NOT_SELECTED":
            facts, facts_cut = await _rows(s, db.facts,
                ["id","source_kind","source_run_id","superseded_by","invalid_at","expired_at","created_at","confidence"], where)
            ids = {f["id"] for f in facts}
            for f in facts:
                f["source_kind"] = _enum(f["source_kind"], {"human","run","intervention","note"})
                f["confidence"] = _finite(f["confidence"])
                if f["confidence"] is not None and f["confidence"]>1:
                    f["confidence"] = None
                # No foreign fact ID leaked via a superseding relationship.
                f["superseded_outside_window"] = f["superseded_by"] is not None and f["superseded_by"] not in ids
                if f["superseded_by"] not in ids:
                    f["superseded_by"] = None
        columns = [db.skill_evaluations.c[k] for k in
            ("id","skill_id","baseline_version_id","candidate_version_id","status","verdict","applied","approval_id","updated_at")]
        metric_names = []
        for side in ("baseline", "candidate"):
            for metric in ("runs", "success_rate", "avg_duration_ms"):
                key = f"{side}_{metric}"
                metric_names.append(key)
                # Select numeric leaves only. Never load arbitrary evaluation text.
                leaf = db.skill_evaluations.c.metrics[side][metric]
                columns.append(leaf.label(key))
                # SQLite json_extract converts true to integer 1. Inspect the
                # original JSON type too, rather than certifying a boolean as
                # one measured run. PostgreSQL JSON preserves its own type.
                dialect = s.bind.dialect.name
                json_type = (sa.func.json_type(db.skill_evaluations.c.metrics, f"$.{side}.{metric}")
                             if dialect == "sqlite" else sa.func.json_typeof(leaf)
                             if dialect == "postgresql" else sa.literal(None))
                columns.append(json_type.label("_type_" + key))
        eval_rows = [dict(r) for r in (await s.execute(sa.select(*columns)
            .order_by(db.skill_evaluations.c.id.desc()).limit(51))).mappings()]
        evaluations, eval_cut = eval_rows[:50], len(eval_rows)>50
        for e in evaluations:
            for key in metric_names:
                e[key] = _finite(e[key]) if e.pop("_type_" + key) in {"integer", "real", "number"} else None
                if e[key] is not None and (key.endswith("_success_rate") and e[key]>1 or key.endswith("_runs") and int(e[key])!=e[key]):
                    e[key] = None
            e["status"] = _enum(e["status"], {"collecting","decided"})
            e["verdict"] = _enum(e["verdict"], {"promote","reject","human_review"})
        counts = dict(Counter(t["status"] for t in tasks))
    return {
        "schema_version": 1, "observed_at": datetime.now(timezone.utc).isoformat(),
        "source": "CANONICAL_BCC_DATABASE", "read_only": True,
        "tasks": tasks, "window_limit": 50, "truncated": truncation,
        "brain": {"task_id": task_id, "runs": runs, "tool_calls": calls, "approvals": approvals,
                  "events": records, "recorded_verifications": verifications, "independent_verification": "NOT_CAPTURED_BY_THIS_VIEW",
                  "note": "Execution and ledger values are recorded state, not a live-model certificate or hidden chain-of-thought."},
        "memory": {"scope": memory_scope, "project_id": project_id, "task_id": task_id if project_id is None else None,
                   "facts": facts, "truncated": facts_cut, "content_exposed": False,
                   "note": "Metadata/lineage map only. No automatic scan, no embeddings or private text displayed."},
        "learning": {"scope": "INSTANCE_EVALUATION_METADATA", "evaluations": evaluations, "truncated": eval_cut,
                     "view_can_promote": False, "note": "A proposal or promote verdict is not an applied skill."},
        "analysis": {"scope": "LATEST_50_TASK_RECORDS", "counts": counts,
                     "observations": [
                         {"code": "FAILED_TASKS", "count": counts.get("failed", 0)},
                         {"code": "WAITING_APPROVAL", "count": counts.get("waiting_approval", 0)},
                         {"code": "COMPLETED_WITHOUT_VERIFICATION_RECORD_IN_WINDOW", "count": sum(r["status"]=="completed" and r["id"] not in {v["run_id"] for v in verifications} for r in runs)},
                         {"code": "TOOL_ERROR_IN_SELECTED_WINDOW", "count": sum(c["status"] in {"error","timeout","interrupted"} for c in calls)}],
                     "automatic_fix": False, "intelligence_score": None},
    }


@router.get("/observatory/snapshot")
async def get_snapshot(request: Request, task_id: int | None = Query(None, ge=1),
                       project_id: str | None = Query(None, min_length=1, max_length=96, pattern=r"^[A-Za-z0-9][A-Za-z0-9_-]*$")):
    try:
        async with asyncio.timeout(5):
            return await snapshot(request.app.state.svc, task_id, project_id)
    except (TimeoutError, sa.exc.SQLAlchemyError):
        raise HTTPException(503, "Observatory data unavailable; retry after database recovery") from None


@router.get("/observatory/routes")
async def get_routes(request: Request):
    """Declared FastAPI contracts, not a claimed distributed trace or DB span."""
    # Reuse the canonical traversal: recent FastAPI keeps included routers
    # lazy rather than flattening app.routes. Do not maintain a second walker.
    from .command_bar import _walk_routes
    routes = []
    for route in _walk_routes(request.app.routes):
        path = getattr(route, "path", "")
        if not path.startswith("/api/"):
            continue
        endpoint = getattr(route, "endpoint", None)
        routes.append({"path": path, "methods": sorted(getattr(route, "methods", []) or []),
                       "handler": (getattr(endpoint, "__module__", "") + "." + getattr(endpoint, "__name__", "")),
                       "evidence": "DECLARED_ROUTE_NOT_EXECUTION"})
    return {"routes": routes[:2000], "truncated": len(routes)>2000, "source": "RUNNING_FASTAPI_ROUTER"}

FEATURE = Feature(name="observatory", router=router)
